proposals/nnnn-function-body-macros.md

```diff
@@ -0,0 +1,338 @@
+
+```swift
+@Logged var area: Double {
+  return length * width
+}
+```
+
+## Detailed design
+
+### Declaring function body macros
+
+Function body macros are declared with the `body` or `preamble` role, which indicate that they can be attached to any kind of function, and can produce the contents of a function body. For example, here are declarations for the macros used above:
+
+```swift
+@attached(body) macro Remote() = #externalMacro(...)
+
+@attached(preamble) macro Logged() = #externalMacro(...)
+
+@attached(body) macro AssumeMainActor() = #externalMacro(...)
+```
+
+Like other attached macros, function body macros have no return type. 
+
+### Implementing function body macros
+
+Body macros are implemented with a type that conforms to the `BodyMacro` protocol:
+
+```swift
+/// Describes a macro that can create the body for a function.
+public protocol BodyMacro: AttachedMacro {
+  /// Expand a macro described by the given custom attribute and
+  /// attached to the given declaration and evaluated within a
+  /// particular expansion context.
+  ///
+  /// The macro expansion introduces code block items that will become the body for the
+  /// given function. Any existing body will be implicitly ignored.
+  static func expansion(
+    of node: AttributeSyntax,
+    providingBodyFor declaration: some DeclSyntaxProtocol & WithOptionalCodeBlockSyntax,
+    in context: some MacroExpansionContext
+  ) throws -> [CodeBlockItemSyntax]
+}
+```
+
+That function may have a function body, which will be replaced by the code items produced from the macro implementation.
+
+Preamble macros are implemented with a type that conforms to the `PreambleMacro` protocol:
+
+```swift
+/// Describes a macro that can introduce "preamble" code into an existing
+/// function body.
+public protocol PreambleMacro: AttachedMacro {
+  /// Expand a macro described by the given custom attribute and
+  /// attached to the given declaration and evaluated within a
+  /// particular expansion context.
+  ///
+  /// The macro expansion can introduce code items that form a preamble to
+  /// the body of the given function. The code items produced by this macro
+  /// expansion will be inserted at the beginning of the function body.
+  static func expansion(
+    of node: AttributeSyntax,
+    providingPreambleFor declaration: some DeclSyntaxProtocol & WithOptionalCodeBlockSyntax,
+    in context: some MacroExpansionContext
+  ) throws -> [CodeBlockItemSyntax]
+}
+```
+
+Preamble macros don't provide a complete function body. Rather, they provide code items that will be introduced at the beginning of the existing function body. Preamble macros are useful for injecting code without changing anything else about the function body, for example to introduce logging or simple tracing facilities. The `defer` statement can be used to trigger code that will run once the function returns.
```

Comment:
There was an interesting question raised [here](https://forums.swift.org/t/function-body-macros/66471/34) which I don't think has been addressed. If a function body pre-expansion is a single-expression implicit return, how do function body macros interact with it?

```swift
@SomeMacro
func f(a: Int, b: Int) {
  a + b
}
```

* If this is a **preamble** macro, the macro doesn't have an opportunity to update the `a + b` expression. Is there a special carve-out for this to still work as an implicit `return` or would that now fail?
* If this is a **body** macro, is it the responsibility of that macro to transform the `expr` `CodeBlockItem` into an `stmt` item with the equivalent `ReturnStmt`?

Reply:
Hmm. I think the expected behavior would be for preamble macros to still perform the implicit `return`. Body macros, on the other hand, would be responsible for adding the `return`.